### ISSUE TITLE
Memory leak in the TDVP

### ISSUE THREAD

**Issue by tszoldra** [markusschmitt/vmc_jax#78]
Hi,
I've found a potential memory leak after some of my long TDVP runs unexpectedly reported out-of-memory errors. The consumed RAM memory seems to grow very slowly but steadily with time.
The minimal working example is below:

    import psutil
    
    import jax.random
    import jax.numpy as jnp
    
    
    from jVMC.vqs import NQS
    from jVMC.nets import FFN
    from jVMC.sampler import MCSampler, propose_spin_flip
    from jVMC.util.tdvp import TDVP
    from jVMC.util.stepper import Euler
    import jVMC.operator as op
    from jVMC.global_defs import device_count
    
    
    
    if __name__ == '__main__':
    
        L = 100
        key = jax.random.PRNGKey(1000)
    
        net = FFN()
        sample_shape = (L,)
        s = jnp.ones((device_count(), 1, *sample_shape))
    
        psi = NQS(net, seed=1234, batchSize=1024)
        psi(s)  # init
    
        hamiltonian = op.BranchFreeOperator()
        for i in range(L):
            hamiltonian.add(op.scal_opstr(1.0, (op.Sx(i),)))
    
        sampler = MCSampler(psi, sample_shape, key, numSamples=1024,
                            updateProposer=propose_spin_flip,
                            sweepSteps=L,
                            numChains=16,
                            thermalizationSweeps=25)
    
        tdvpEquation = TDVP(sampler, diagonalizeOnDevice=True)
        stepper = Euler(timeStep=1e-2)
    
        process = psutil.Process()
        for i in range(1000):
            dp, dt = stepper.step(0, tdvpEquation, psi.get_parameters(),
                                  hamiltonian=hamiltonian, psi=psi,
                                  numSamples=None)
            psi.set_parameters(dp)
    
            if i % 10 == 0:
                ram_usage_mb = process.memory_info().rss / 1024 ** 2
                print(f'{i=} -- RAM [MB]={ram_usage_mb: .2f}')

The code is supposed to evolve the randomly initialized state for 1000 timesteps, and should definitely saturate in the consumed memory. The output of the code shows a steady increase of the total memory used by the process at a rate of few megabytes per 10 iterations:

    i=0 -- RAM [MB]= 1214.50
    i=10 -- RAM [MB]= 1231.77
    i=20 -- RAM [MB]= 1238.03
    i=30 -- RAM [MB]= 1242.71
    i=40 -- RAM [MB]= 1244.62
    i=50 -- RAM [MB]= 1251.53
    i=60 -- RAM [MB]= 1256.61
    i=70 -- RAM [MB]= 1263.08
    i=80 -- RAM [MB]= 1267.17
    i=90 -- RAM [MB]= 1270.91
    i=100 -- RAM [MB]= 1276.10
    i=110 -- RAM [MB]= 1280.11
    i=120 -- RAM [MB]= 1286.55
    i=130 -- RAM [MB]= 1290.87
    i=140 -- RAM [MB]= 1296.86
    i=150 -- RAM [MB]= 1304.73
    i=160 -- RAM [MB]= 1305.80
    i=170 -- RAM [MB]= 1308.84
    i=180 -- RAM [MB]= 1311.93
    i=190 -- RAM [MB]= 1318.57
    i=200 -- RAM [MB]= 1323.18
    i=210 -- RAM [MB]= 1324.70
    i=220 -- RAM [MB]= 1326.54
    i=230 -- RAM [MB]= 1327.85
    i=240 -- RAM [MB]= 1331.09
    i=250 -- RAM [MB]= 1333.82
    i=260 -- RAM [MB]= 1335.04
    i=270 -- RAM [MB]= 1340.18
    i=280 -- RAM [MB]= 1343.83
    i=290 -- RAM [MB]= 1345.96
    i=300 -- RAM [MB]= 1348.96
    i=310 -- RAM [MB]= 1350.68
    i=320 -- RAM [MB]= 1351.29
    i=330 -- RAM [MB]= 1351.88
    i=340 -- RAM [MB]= 1353.76
    i=350 -- RAM [MB]= 1355.43
    i=360 -- RAM [MB]= 1357.55
    i=370 -- RAM [MB]= 1360.71
    i=380 -- RAM [MB]= 1362.68
    i=390 -- RAM [MB]= 1366.70
    i=400 -- RAM [MB]= 1369.30
    i=410 -- RAM [MB]= 1372.06
    i=420 -- RAM [MB]= 1375.64
    i=430 -- RAM [MB]= 1378.47
    i=440 -- RAM [MB]= 1379.09
    i=450 -- RAM [MB]= 1383.06
    i=460 -- RAM [MB]= 1385.54
    i=470 -- RAM [MB]= 1386.68
    i=480 -- RAM [MB]= 1390.14
    i=490 -- RAM [MB]= 1395.88
    i=500 -- RAM [MB]= 1400.59
    i=510 -- RAM [MB]= 1409.31
    i=520 -- RAM [MB]= 1411.14
    i=530 -- RAM [MB]= 1413.51
    i=540 -- RAM [MB]= 1414.63
    i=550 -- RAM [MB]= 1417.09
    i=560 -- RAM [MB]= 1418.11
    i=570 -- RAM [MB]= 1420.41
    i=580 -- RAM [MB]= 1421.75
    i=590 -- RAM [MB]= 1423.54
    i=600 -- RAM [MB]= 1424.79
    i=610 -- RAM [MB]= 1427.93
    i=620 -- RAM [MB]= 1432.79
    i=630 -- RAM [MB]= 1435.91
    i=640 -- RAM [MB]= 1438.34
    i=650 -- RAM [MB]= 1442.00
    i=660 -- RAM [MB]= 1445.05
    i=670 -- RAM [MB]= 1447.55
    i=680 -- RAM [MB]= 1448.70
    i=690 -- RAM [MB]= 1450.34
    i=700 -- RAM [MB]= 1452.84
    i=710 -- RAM [MB]= 1453.95
    i=720 -- RAM [MB]= 1455.76
    i=730 -- RAM [MB]= 1458.64
    i=740 -- RAM [MB]= 1461.31
    i=750 -- RAM [MB]= 1462.48
    i=760 -- RAM [MB]= 1463.76
    i=770 -- RAM [MB]= 1464.86
    i=780 -- RAM [MB]= 1465.51
    i=790 -- RAM [MB]= 1466.61
    i=800 -- RAM [MB]= 1471.58
    i=810 -- RAM [MB]= 1475.31
    i=820 -- RAM [MB]= 1478.77
    i=830 -- RAM [MB]= 1480.80
    i=840 -- RAM [MB]= 1483.38
    i=850 -- RAM [MB]= 1486.15
    i=860 -- RAM [MB]= 1487.84
    i=870 -- RAM [MB]= 1489.71
    i=880 -- RAM [MB]= 1491.89
    i=890 -- RAM [MB]= 1493.16
    i=900 -- RAM [MB]= 1493.73
    i=910 -- RAM [MB]= 1493.79
    i=920 -- RAM [MB]= 1496.02
    i=930 -- RAM [MB]= 1498.51
    i=940 -- RAM [MB]= 1501.33
    i=950 -- RAM [MB]= 1502.48
    i=960 -- RAM [MB]= 1505.44
    i=970 -- RAM [MB]= 1507.34
    i=980 -- RAM [MB]= 1508.36
    i=990 -- RAM [MB]= 1510.28

The test was done with jVMC==1.5.0 and jax==0.4.31, CUDA==12.2 and took ~ 15 minutes on Nvidia RTX 3080 GPU.

I am currently investigating the issue, so far without success.

**Comment by tszoldra** [comment]
The same type of behavior is observed when MinSR is used instead of the TDVP:

    ...
    #tdvpEquation = TDVP(sampler, diagonalizeOnDevice=False)
    tdvpEquation = MinSR(sampler, makeReal='real')    
    ...

Output:

    i=0 -- RAM [MB]= 1203.07
    i=10 -- RAM [MB]= 1267.36
    i=20 -- RAM [MB]= 1271.41
    i=30 -- RAM [MB]= 1275.46
    i=40 -- RAM [MB]= 1280.84
    i=50 -- RAM [MB]= 1283.89
    i=60 -- RAM [MB]= 1287.29
    i=70 -- RAM [MB]= 1290.64
    i=80 -- RAM [MB]= 1292.99
    i=90 -- RAM [MB]= 1296.53
    i=100 -- RAM [MB]= 1299.89
    ....
    i=970 -- RAM [MB]= 1546.75
    i=980 -- RAM [MB]= 1549.11
    i=990 -- RAM [MB]= 1552.60
